app/src/main/java/com/keylesspalace/tusky/util/SaveTootHelper.java

```diff
@@ -0,0 +1,188 @@
+            return null;
+        }
+        ContentResolver contentResolver = context.getContentResolver();
+        ArrayList<File> filesSoFar = new ArrayList<>();
+        ArrayList<String> results = new ArrayList<>();
+        for (ComposeActivity.QueuedMedia item : mediaQueued) {
+            /* If the media was already saved in a previous draft, there's no need to save another
+             * copy, just add the existing URI to the results. */
+            if (existingUris != null) {
+                String uri = item.uri.toString();
+                int index = existingUris.indexOf(uri);
+                if (index != -1) {
+                    results.add(uri);
+                    continue;
+                }
+            }
+            // Otherwise, save the media.
+            File directory;
+            switch (item.type) {
+                default:
+                case IMAGE:
+                    directory = imageDirectory;
+                    break;
+                case VIDEO:
+                    directory = videoDirectory;
+                    break;
+            }
+            String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US)
+                    .format(new Date());
+            String mimeType = contentResolver.getType(item.uri);
+            MimeTypeMap map = MimeTypeMap.getSingleton();
+            String fileExtension = map.getExtensionFromMimeType(mimeType);
+            String filename = String.format("Tusky_Draft_Media_%s.%s", timeStamp, fileExtension);
+            File file = new File(directory, filename);
+            filesSoFar.add(file);
+            boolean copied = IOUtils.copyToFile(contentResolver, item.uri, file);
+            if (!copied) {
+                /* If any media files were created in prior iterations, delete those before
+                 * returning. */
+                for (File earlierFile : filesSoFar) {
+                    boolean deleted = earlierFile.delete();
+                    if (!deleted) {
+                        Log.i(TAG, "Could not delete the file " + earlierFile.toString());
+                    }
+                }
+                return null;
+            }
+            Uri uri = FileProvider.getUriForFile(context, BuildConfig.APPLICATION_ID+".fileprovider", file);
+            results.add(uri.toString());
+        }
+        return results;
+    }
+
+    private void deleteMedia(List<String> mediaUris) {
+        for (String uriString : mediaUris) {
+            Uri uri = Uri.parse(uriString);
+            if (context.getContentResolver().delete(uri, null, null) == 0) {
+                Log.e(TAG, String.format("Did not delete file %s.", uriString));
+            }
+        }
+    }
+
+    /**
+     * A∖B={x∈A|x∉B}
+     *
+     * @return all elements of set A that are not in set B.
+     */
+    private static List<String> setDifference(List<String> a, List<String> b) {
+        List<String> c = new ArrayList<>();
+        for (String s : a) {
```

Comment:
This is some expensive stuff, like O(N^2)... it's probably doesn't matter here and I'm not sure that sorting lists in prior is cheaper

Reply:
Thats no new code, its cut from `ComposeActivity`. And it is only called on attachment lists where n is at most 4.